### PR TITLE
[WIP] POC: mempool, swap data ptr of arrays, etc.

### PR DESCRIPTION
**Pull request by KDr2**
In the initial commit:
1. constructing arrays from a pre-allocated memory pool
2. swap data ptr between these arrays

### CODE REVIEW

File: src/mempool.jl [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    # the stack-allocated buffer will be released.
+    ptr = unsafe_convert(Ptr{UInt8}, global_buf) + 0
+    a1 = unsafe_wrap(Array{UInt8, 2}, ptr, (2, 2))
+    println("Hi, I am from Julia")
+    println(typeof(a1))
+    println(stack_buffer)
+    dest = Core.bitcast(Ptr{Cvoid}, stack_buffer)
+    @show dest
+
+    a2 = ccall(copy_to_stack_array, Any, (Any, Ptr{Cvoid}), a1, dest)
```
Comment (yebai) [inline src/mempool.jl:30]: Maybe consider wrapping L22-L30 into a new function?

File: src/mempool.jl [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+function entrance_t(stack_buffer)
+    # my caller is `lt_stack_alloc`, when I return,
+    # the stack-allocated buffer will be released.
+    ptr = unsafe_convert(Ptr{UInt8}, global_buf) + 0
+    a1 = unsafe_wrap(Array{UInt8, 2}, ptr, (2, 2))
+    println("Hi, I am from Julia")
+    println(typeof(a1))
+    println(stack_buffer)
+    dest = Core.bitcast(Ptr{Cvoid}, stack_buffer)
```
Comment (yebai) [inline src/mempool.jl:57]: Maybe move `Core.bitcast(Ptr{Cvoid}, stack_buffer)` into `lt_stack_alloc`, then we can definie `entrance_t(p::Ptr{Cvoid}` instead of ``entrance_t(p::Any)`? It slightly improves clarity and safety I think.

File: src/mempool.jl [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    a2 = ccall(copy_to_stack_array, Any, (Any, Ptr{Cvoid}), a1, dest)
+    @show a2, a2[1], a2[2]
+    istaskdone(t1) || yieldto(t1)
+    @show a2, a2[1], a2[2]
+    istaskdone(t1) || yieldto(t1)
+    @show a2, a2[1], a2[2]
+    istaskdone(t1) || yieldto(t1)
+end
+
+function test_t()
```
Comment (yebai) [inline src/mempool.jl:68]: Maybe create a new function since this is needed everywhere we want to create a new task, e.g. 

```julia
call_with_stackbuff(entrance::Function) = ccall(stack_alloc, Any, (Any, UInt), entrance, 1024)

test_t() = call_with_stackbuff(entrance_t)
```

**Comment by yebai** [comment]
@KDr2 FYI, it appears we can get pointer for arrays using:

```julia
a = [1,2,3]
ap = Base.unsafe_convert(Ptr{Clonglong}, a)
a2 = unsafe_wrap(Vector{Int64}, ap, 3, own=false)

```

Ref: https://stackoverflow.com/a/57946260

File: src/mempool.jl [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+test()
+
+# Task test
+println("============= task test =")
+
+tr = Ref{Task}()
+t1 = @task for _ in 1:4
+    @show "---"
+    yieldto(tr[])
```
[the remark this answers is not in the document]
Comment (KDr2) [reply src/mempool.jl:46]: It is assigned at around line 75.

**Comment by KDr2** [comment]
About `alloca`: `alloca` is a c macro defined as `__builtin_alloca` which is an inline function, so both of them are not callable via `ccall` in Julia because the symbol is not exported.

File: src/mempool.c [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+JL_DLLEXPORT jl_value_t * lt_stack_alloc(jl_value_t *fun, size_t buf_size)
+{
+    void *buf = alloca(buf_size);
+    jl_function_t* func = (jl_function_t*)fun;
+    jl_value_t *argument = jl_box_int64((int64_t)buf);
+    jl_value_t *ret = jl_call1(func, argument);
+    return ret;
+}
+
+JL_DLLEXPORT jl_array_t* lt_copy_to_stack_array(jl_array_t *ary, void *dest)
```
Comment (KDr2) [inline src/mempool.c:62]: After trying a little, I think implementing this function in Julia is not a good idea:
- We always at least need a c function that converts a Julia object reference to a memory address, so `ccall` a c function is not avoidable.
- There will be many "magic numbers" if we implement it in Julia, say:
  ```
  ccall(:memcpy, Ptr{Cvoid}, (Ptr{Cvoid}, Ptr{Cvoid}, UInt), 
           dest, obj_addr - M1, M2);
  ```
  where `M1` is `sizeof(jl_taggedvalue_t)` and M2 is `sizeof(jl_taggedvalue_t) + sizeof(jl_array_t)`.
  Here `M1` and `M2` can be calculated in compiling time in C code, but when in Julia, they cannot be calculated so that we must use magic numbers, and even worse, M2 will vary according to the type of the object we are copying.

Comment (yebai) [reply src/mempool.c:62]: It sounds good to focus on making sure this works, and figure out all the edge cases for now!

Comment (KDr2) [reply src/mempool.c:62]: Error message:
```
============= libtask test =
Hi, I am from Julia
Array{UInt8,2}
139977267218336
dest = Ptr{Nothing} @0x00007f4eff4a07a0
consume(ctask) = 0x01
consume(ctask) = 0x02

signal (11): Segmentation fault
in expression starting at /data/zhuoql/Work/julia/Libtask.jl/src/mempool.jl:119
page_metadata at /buildworker/worker/package_linux64/build/src/gc.h:448 [inlined]
gc_setmark_pool at /buildworker/worker/package_linux64/build/src/gc.c:751 [inlined]
gc_mark_loop at /buildworker/worker/package_linux64/build/src/gc.c:2402
_jl_gc_collect at /buildworker/worker/package_linux64/build/src/gc.c:2902
jl_gc_collect at /buildworker/worker/package_linux64/build/src/gc.c:3108
maybe_collect at /buildworker/worker/package_linux64/build/src/gc.c:827 [inlined]
jl_gc_pool_alloc at /buildworker/worker/package_linux64/build/src/gc.c:1142
jl_gc_alloc_ at /buildworker/worker/package_linux64/build/src/julia_internal.h:277 [inlined]
jl_gc_alloc at /buildworker/worker/package_linux64/build/src/gc.c:3150
_new_array_ at /buildworker/worker/package_linux64/build/src/array.c:94 [inlined]
_new_array at /buildworker/worker/package_linux64/build/src/array.c:162 [inlined]
jl_alloc_array_1d at /buildworker/worker/package_linux64/build/src/array.c:433
Array at ./boot.jl:406 [inlined]
Array at ./boot.jl:415 [inlined]
zeros at ./array.jl:525 [inlined]
zeros at ./array.jl:521 [inlined]
BitSet at ./bitset.jl:18
find_ssavalue_uses at ./compiler/utilities.jl:193
InferenceState at ./compiler/inferencestate.jl:77
InferenceState at ./compiler/inferencestate.jl:121
typeinf_ext at ./compiler/typeinfer.jl:568
typeinf_ext at ./compiler/typeinfer.jl:601
jfptr_typeinf_ext_23344.clone_1 at /data/zhuoql/programs/julia-1.5.3/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2214 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2398
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1690 [inlined]
jl_type_infer at /buildworker/worker/package_linux64/build/src/gf.c:296
jl_generate_fptr at /buildworker/worker/package_linux64/build/src/jitlayers.cpp:290
jl_compile_method_internal at /buildworker/worker/package_linux64/build/src/gf.c:1964
jl_compile_method_internal at /buildworker/worker/package_linux64/build/src/gf.c:1919 [inlined]
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2224 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2398
copy at /data/zhuoql/.julia/packages/Libtask/wVNyg/src/ctask.jl:99
unknown function (ip: 0x7f4ee6a2949c)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2231 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2398
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1690 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:117
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:206
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:157 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:566
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:660
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:840
jl_parse_eval_all at /buildworker/worker/package_linux64/build/src/ast.c:913
jl_load_rewrite at /buildworker/worker/package_linux64/build/src/toplevel.c:914
include at ./Base.jl:380
include at ./Base.jl:368
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2214 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2398
exec_options at ./client.jl:296
_start at ./client.jl:506
jfptr__start_53898.clone_1 at /data/zhuoql/programs/julia-1.5.3/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2214 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2398
unknown function (ip: 0x401931)
unknown function (ip: 0x401533)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x4015d4)
Allocations: 1649084 (Pool: 1648560; Big: 524); GC: 1
Segmentation fault (core dumped)
```

**Comment by KDr2** [comment]
Here is what happens when we copy a mutable object (array as an example) onto the stack:

* in C code, `alloca` memory on the stack, copy an object to it by calling `memcpy`
*  in Julia code, we hold a reference to the object we just copied onto the stack: `a = copy_to_stack(...)`
* When the Julia code is compiled to LLVM IR, the compiler will insert some GC relevant code about the reference `a`, like put it into a `gcframe` or similar GC relevant data structure, which tells the GC subsystem to manage `a`, then the GC subsystem will treat the memory where `a` is placed as a GC-managed memory pool, that crashes the process while GC occurring. 
* Here what we expect is that `a` is a stack-allocated and GC won't manage it. So we should figure out a way to tell the compiler not to insert the GC relevant code about `a`, that is impossible without altering the compiler. I tried using a `WeakRef` to hold `a`, it reduces the crashes but doesn't reduce all of them: we eventually will get `a` out from the weak reference if we want to access `a`.

So, to conclude, I think we hardly can implement a robust, maintainable data-share/copy way between tasks in this approach.

**Comment by yebai** [comment]
Many thanks, @KDr2 for the exploration. As discussed, let's close this in favour of https://github.com/TuringLang/Libtask.jl/pull/82
